src/utils/useOnDataChanged.ts

```diff
@@ -0,0 +1,39 @@
+import { UseInfiniteQueryResult, UseQueryResult } from '@tanstack/react-query';
+import { useEffect } from 'react';
+
+type OnChangedCallBack<T> = { onSuccess: (newData: T | undefined) => void };
+
+// Use function overload to accept UseInfiniteQueryResult or UseQueryResult and return the correct type, depending on the input's type.
+export function useOnDataChanged<T>(
+  query: UseInfiniteQueryResult<T>,
+  { onSuccess }: OnChangedCallBack<T>,
+): UseInfiniteQueryResult<T>;
+
+export function useOnDataChanged<T>(
+  query: UseQueryResult<T>,
+  { onSuccess }: OnChangedCallBack<T>,
+): UseQueryResult<T>;
+
+/**
+ * Tanstack Query 5 removed onSuccess, onError and onSettled in useQuery.
+ * Read https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose for more information.
+ * This function allows to call a callback when the useQuery succeed.
+ *
+ * @param query The useQuery result to observe.
+ * @param onDataChanged The callback to call once the query is a success.
+ * @returns The UseQueryResult to faciliate the usage.
+ */
+export function useOnDataChanged<T>(
+  query: UseQueryResult<T> | UseInfiniteQueryResult<T>,
+  { onSuccess }: OnChangedCallBack<T>,
+): UseQueryResult<T> | UseInfiniteQueryResult<T> {
+  const { data } = query;
+
+  useEffect(() => {
+    if (data && !query.error) {
+      onSuccess(data);
+    }
+  }, [data]);
+
+  return query;
+}
```

Comment:
While I see why you implemented such a function, I think we will end-up with duplicated state if we use the hook more than once in the app as explained in the article you linked, see the [relevant section](https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose#a-bad-api) about non-deduplicated call to onError.